Gems/MotionMatching/Code/Source/MotionMatchingInstance.cpp

```diff
@@ -485,20 +488,35 @@ namespace EMotionFX::MotionMatching
         const FeatureSchema& featureSchema = m_data->GetFeatureSchema();
         const FeatureTrajectory* trajectoryFeature = m_cachedTrajectoryFeature;
 
-        // 1. Broad-phase search using KD-tree
-        if (mm_useKdTree)
+        // 1. Build query vector
         {
+            AZ_PROFILE_SCOPE(Animation, "MM::BuildQueryVector");
+
             // Build the input query features that will be compared to every entry in the feature database in the motion matching search.
+            AZ_Assert(m_queryVector.size() == aznumeric_cast<size_t>(m_data->GetFeatureMatrix().cols()),
+                "The query vector should have the same number of elements as the feature matrix has columns.");
+            for (Feature* feature : featureSchema.GetFeatures())
+            {
+                feature->FillQueryVector(m_queryVector, queryVectorContext);
+            }
+        }
+
+        // 2. Broad-phase search using KD-tree
+        if (mm_useKdTree)
+        {
+            AZ_PROFILE_SCOPE(Animation, "MM::BroadPhaseKDTree");
+
+            // For the kd-tree
             size_t startOffset = 0;
             for (Feature* feature : m_data->GetFeaturesInKdTree())
             {
-                feature->FillQueryFeatureValues(startOffset, m_queryFeatureValues, context);
+                memcpy(&m_kdTreeQueryVector[startOffset], &m_queryVector[feature->GetColumnOffset()], feature->GetNumDimensions() * sizeof(float));
```

Comment:
Just checked and AZStd::copy does it element by element using iterators. I want to copy a block of memory for performance reasons.

Reply:
It's possible/likely the internal implementations will actually fallback to a `memcpy` for pod types using some `constexpr` magic most likely, but it might not in _all_ cases (this might actually be something interesting to write a quick benchmark for in [quick-bench](https://quick-bench.com/) 😛 to see if there's a noticeable difference.